espnet2/tasks/abs_task.py

```diff
@@ -620,88 +686,66 @@ def main(cls, args: argparse.Namespace = None, cmd: Sequence[str] = None) -> Non
             cls.print_config()
             sys.exit(0)
         cls.check_required_command_args(args)
-        logging.basicConfig(
-            level=args.log_level,
-            format="%(asctime)s (%(module)s:%(lineno)d) " "%(levelname)s: %(message)s",
-        )
-
-        # 1. Set random-seed
-        random.seed(args.seed)
-        np.random.seed(args.seed)
-        torch.random.manual_seed(args.seed)
 
-        # 2. Build train-data-iterator
-        if args.train_dtype in ("float32", "O0", "O1", "O2", "O3"):
-            dtype = "float32"
+        # 0. Init distributed process
+        distributed_option = build_dataclass(DistributedOption, args)
+        distributed_option.init()
+        if not distributed_option.distributed or distributed_option.dist_rank == 0:
+            logging.basicConfig(
+                level=args.log_level,
+                format="%(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s",
+            )
         else:
-            dtype = args.train_dtype
-        train_dataset = ESPnetDataset(
-            args.train_data_path_and_name_and_type,
-            float_dtype=dtype,
-            preprocess=cls.build_preprocess_fn(args, True),
-        )
-        cls.check_task_requirements(train_dataset, args.allow_variable_data_keys)
-        train_batch_sampler = build_batch_sampler(
-            type=args.batch_type,
+            logging.basicConfig(
+                level="ERROR",
+                format="%(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s",
+            )
+
+        # 1. Set random-seed
+        set_all_random_seed(args.seed)
+        torch.backends.cudnn.enabled = args.cudnn_enabled
+        torch.backends.cudnn.benchmark = args.cudnn_benchmark
+        torch.backends.cudnn.deterministic = args.cudnn_deterministic
+
+        # 2. Build iterator factories
+        train_iter_factory = cls.build_iter_factory(
+            args=args,
```

Comment:
args is a little bit confusing since this function has other arguments.

Reply:
I agree this is bad way, but I have no idea about this.

My intention of giving args is  to make `build_iter_factory` abstracted and enable to be overwritten.

```python
class SomeTask(AbsTask):
    def build_iter_factory(cls, args, ...):
       ...
```

Basically, the task class assumed same data-iterator among many tasks, but
some tasks, e.g. LM training with Truncated BPTT,  require splitting the sequence into pieces and this iterator doesn't support such flexibility. I'd like to add such features.

If args is converted explicit names, we can't give additional arguments for such unseen iterator.

Reply:
Well, I need more time to think about this, but in this PR, I'd like to keep this shape.

Reply:
I see. OK.